JSONize plant growth stages

RenechCDDA opened the pull request:
#### Summary
Features "Plant-specific growth stages for agriculture"

#### Purpose of change
Clean up some of the TODO/FIXMEs I've written in the past.

Steps towards expanding agriculture so it isn't so binary

#### Describe the solution
Replace the "grow" duration with a "growth_stages" duration defining a (generic!) flag for a growth stage and a duration.

This allows plants which stay in seedling form for most of their lives, or have unusually short/long harvest periods, etc.

To accommodate this a new GROWTH_SEED stage was added, to define how long they should stay in that form.

A dumb find and replace was used to replace old "grow" values with "growth_stages". When the existing growth duration was not cleanly divisible by 4, one extra day was added to earlier stages until it summed to the previous total. (e.g. a duration of 30 became 8+8+7+7)

<details>
<summary>find and replace details:</summary>

"grow": "280 days"
"growth_stages": { "GROWTH_SEED": "70 days", "GROWTH_SEEDLING": "70 days", "GROWTH_MATURE": "70 days", "GROWTH_HARVEST": "70 days" }

"grow": "130 days"
"growth_stages": { "GROWTH_SEED": "33 days", "GROWTH_SEEDLING": "33 days", "GROWTH_MATURE": "32 days", "GROWTH_HARVEST": "32 days" }

"grow": "120 days"
"growth_stages": { "GROWTH_SEED": "30 days", "GROWTH_SEEDLING": "30 days", "GROWTH_MATURE": "30 days", "GROWTH_HARVEST": "30 days" }

"grow": "95 days"
"growth_stages": { "GROWTH_SEED": "24 days", "GROWTH_SEEDLING": "24 days", "GROWTH_MATURE": "24 days", "GROWTH_HARVEST": "23 days" }

"grow": "91 days"
"growth_stages": { "GROWTH_SEED": "23 days", "GROWTH_SEEDLING": "23 days", "GROWTH_MATURE": "23 days", "GROWTH_HARVEST": "22 days" }

"grow": "90 days"
"growth_stages": { "GROWTH_SEED": "23 days", "GROWTH_SEEDLING": "23 days", "GROWTH_MATURE": "22 days", "GROWTH_HARVEST": "22 days" }

"grow": "85 days"
"growth_stages": { "GROWTH_SEED": "22 days", "GROWTH_SEEDLING": "21 days", "GROWTH_MATURE": "21 days", "GROWTH_HARVEST": "21 days" }

"grow": "80 days"
"growth_stages": { "GROWTH_SEED": "20 days", "GROWTH_SEEDLING": "20 days", "GROWTH_MATURE": "20 days", "GROWTH_HARVEST": "20 days" }

"grow": "75 days"
"growth_stages": { "GROWTH_SEED": "19 days", "GROWTH_SEEDLING": "19 days", "GROWTH_MATURE": "19 days", "GROWTH_HARVEST": "18 days" }

"grow": "70 days"
"growth_stages": { "GROWTH_SEED": "19 days", "GROWTH_SEEDLING": "17 days", "GROWTH_MATURE": "17 days", "GROWTH_HARVEST": "17 days" }

"grow": "65 days"
"growth_stages": { "GROWTH_SEED": "17 days", "GROWTH_SEEDLING": "16 days", "GROWTH_MATURE": "16 days", "GROWTH_HARVEST": "16 days" }

"grow": "60 days"
"growth_stages": { "GROWTH_SEED": "15 days", "GROWTH_SEEDLING": "15 days", "GROWTH_MATURE": "15 days", "GROWTH_HARVEST": "15 days" }

"grow": "50 days"
"growth_stages": { "GROWTH_SEED": "13 days", "GROWTH_SEEDLING": "13 days", "GROWTH_MATURE": "12 days", "GROWTH_HARVEST": "12 days" }

"grow": "47 days"
"growth_stages": { "GROWTH_SEED": "12 days", "GROWTH_SEEDLING": "12 days", "GROWTH_MATURE": "12 days", "GROWTH_HARVEST": "11 days" }

"grow": "35 days"
"growth_stages": { "GROWTH_SEED": "9 days", "GROWTH_SEEDLING": "9 days", "GROWTH_MATURE": "9 days", "GROWTH_HARVEST": "8 days" }
</details>


Modify the C++ to work on the seed's actual growth stage durations instead of generic epochs

Remove "grow" field and update docs to explain "growth_stages"

#### Describe alternatives you've considered


#### Testing
Game loads with current changes

New data is on the plants
<img width="405" height="59" alt="image" src="https://github.com/user-attachments/assets/3dd32dcc-4611-4c89-9614-1db691cdb1c6" />

Plants grow according to their new growth schedules. (Basically the same as old ones, currently.)

Plant some seeds, teleport away, advance a year, teleport back. Tile became overgrown.

Plant some seeds, teleport away, advance time for one stage, teleport back. Tile grew one stage.

Added some debug plumbing to DF_MAP to sanity check the stage advances.

#### Additional context
Please do not squash :)

Each commit has a purpose and one of the commits is not authored by me, authorship should be preserved by avoiding squashing.

RenechCDDA committed:
Implement growth_stages data using generic reader, and convert json data

-Add GROWTH_SEED flag for newly planted seeds

ehughsbaird commented on src/itype.h:
```diff
@@ -1161,51 +1161,46 @@ struct islot_bionic {
+        ter_furn_flag required_terrain_flag = ter_furn_flag::TFLAG_PLANTABLE;
+        islot_seed() = default;
+
+        std::map<flag_id, time_duration> get_growth_stages() const;
```
Return a const reference, there's no need to copy here.

PatrikLundell reviewed (commented):
Nitpicking: I'd place the GROWTH_SEED flags last, to match the placement of GROWTH_SEEDLING.

JSON_FLAGS.md: GROWTH_SEEDL -> GROWTH_SEED.

Comment on the Testing image:
"GROWTH_MATURE and 2 weeks and 6 days" is an odd phrase. My guess is that the first "and" probably should be either "lasting for" or "remaining". Can't see any corresponding code change, though.

RenechCDDA commented:
> "GROWTH_MATURE and 2 weeks and 6 days" is an odd phrase. My guess is that the first "and" probably should be either "lasting for" or "remaining". Can't see any corresponding code change, though.

Oh yeah, it was just some super basic debug output so I could be sure that the data was on the item. Not meant to be exposed to the player right now, this will probably go into iteminfo in some way. (I haven't decided)

PatrikLundell commented:
Ah, that explains why the code for it wasn't included.

My understanding of autumn planted crops is that they start to grow during the autumn, with the important part being the establishment of roots. The overground parts behave like other grasses. All this means that the plants get a head start in spring.

The purpose of cover crops is to grow during the autumn and have the roots and above ground parts keep the soil from being exposed (and it doesn't really matter if they're withered or not: they'll still protect the soil), and then the plants are destroyed in spring when plowing for the "real" crops.

I believe nitrogen fixing plants (like peas) as well as rye (which doesn't seem to be used as a regular crop in New England, sadly) are used as cover crops.

Note that I'm not a farmer and have no specific knowledge in the area, though, so take the above with a grain of salt.

RenechCDDA commented:
> My understanding of autumn planted crops is that they start to grow during the autumn, with the important part being the establishment of roots. The overground parts behave like other grasses.

By the same measure we could have them stick in "seedling" or "mature" form for the same effect. All that would matter is the time between two stages is sufficiently long to get past the part of winter where they won't grow. (Of course we could also add things to the 'forward looking' to make it easier for specific plants to grow, but that's out of scope for here)

Anyway the specifics can be hashed out once I finish this up. Mostly just needs some work to handle flag_id.

RenechCDDA commented:
So, I forgot that map does not preserve order of insertion. We need to iterate in order of insertion (that is *very* important) so... vector of pairs instead?

ehughsbaird committed:
Generic reader for vector of pairs

RenechCDDA committed:
Implement handling for variable seed growth stages
-Switch to using a vector of pairs to preserve insertion order (also order of iteration, VERY IMPORTANT!)

-Remove the remnants of "grow", document growth_stages
